### PR TITLE
108424 prohibit change after clear verify

### PR DESCRIPTION
**Pull request by janid1967**
[AB#108424](https://statoildeveloper.visualstudio.com/b6d97460-2e53-47ae-aad4-b41396d90828/_workitems/edit/108424) 

Rules:
image

### CODE REVIEW

File: src/Equinor.ProCoSys.Completion.Command/PunchItemCommands/DeletePunchItem/DeletePunchItemCommandValidator.cs [hunk 1]
```diff
@@ -17,12 +17,17 @@ public DeletePunchItemCommandValidator(IPunchItemValidator punchItemValidator)
             .MustAsync((command, cancellationToken) => NotBeInAClosedProjectForPunchItemAsync(command.PunchItemGuid, cancellationToken))
             .WithMessage("Project is closed!")
             .MustAsync((command, cancellationToken) => BeAnExistingPunchItemAsync(command.PunchItemGuid, cancellationToken))
-            .WithMessage(command => $"Punch item with this guid does not exist! Guid={command.PunchItemGuid}");
+            .WithMessage(command => $"Punch item with this guid does not exist! Guid={command.PunchItemGuid}")
+            .MustAsync((command, cancellationToken) => NotBeClearedAsync(command.PunchItemGuid, cancellationToken))
+            .WithMessage(command => $"Punch item is cleared! Guid={command.PunchItemGuid}");
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.Completion.Command/PunchItemCommands/DeletePunchItem/DeletePunchItemCommandValidator.cs:23]: Will this message be presented to users? Does it provide enough information? Why does it matter that the punch item is cleared. What can be done about it. Who is interested in the Guid?

File: src/Equinor.ProCoSys.Completion.Command/PunchItemCommands/DeletePunchItemAttachment/DeletePunchItemAttachmentCommandValidator.cs [hunk 2]
```diff
@@ -26,7 +26,9 @@ public class DeletePunchItemAttachmentCommandValidator : AbstractValidator<Delet
             .MustAsync((command, cancellationToken) => NotBeInAVoidedTagForPunchItemAsync(command.PunchItemGuid, cancellationToken))
             .WithMessage("Tag owning punch item is voided!")
             .MustAsync((command, cancellationToken) => NotBeVerifiedAsync(command.PunchItemGuid, cancellationToken))
-            .WithMessage(command => $"Punch item attachments can't be changed. The punch item is verified! Guid={command.PunchItemGuid}");
+            .WithMessage(command => $"Punch item attachments can't be changed. The punch item is verified! Guid={command.PunchItemGuid}")
+            .UnlessAsync((command, cancellationToken)
```
Comment (trondlindanger) [inline src/Equinor.ProCoSys.Completion.Command/PunchItemCommands/DeletePunchItemAttachment/DeletePunchItemAttachmentCommandValidator.cs:30]: Does the receiver get enough information to take action?

File: src/tests/Equinor.ProCoSys.Completion.Command.Tests/PunchItemCommands/CreatePunchItemLink/CreatePunchItemLinkCommandValidatorTests.cs [hunk 3]
```diff
@@ -78,4 +78,20 @@ public async Task Validate_ShouldFail_When_ProjectIsClosed()
...
+            .Returns(true);
+
+        // Act
+        var result = await _dut.ValidateAsync(_command);
+
+        // Assert
+        Assert.IsFalse(result.IsValid);
+        Assert.AreEqual(1, result.Errors.Count);
+        Assert.IsTrue(result.Errors[0].ErrorMessage.StartsWith("Punch item is cleared!"));
+    }
```
Comment (trondlindanger) [inline src/tests/Equinor.ProCoSys.Completion.Command.Tests/PunchItemCommands/CreatePunchItemLink/CreatePunchItemLinkCommandValidatorTests.cs:96]: This test should not fail for all actions. Hence the given state it should fail under should be provided in the method name. Eg: Validate_ShouldFail_When_AddLinkForClearedPunchItem? Not clear to me how the action is provided in the selected test method naming convention.

File: src/Equinor.ProCoSys.Completion.Command/PunchItemCommands/DeletePunchItem/DeletePunchItemCommandValidator.cs (hunk 1 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.Command/PunchItemCommands/DeletePunchItem/DeletePunchItemCommandValidator.cs:23]: This is incorporated business rules in Punch domain. Also as is in PCS4.
The ui should disable the delete-button (and other illegal actions when cleared), so the endpoint never be called when signed.
These rules in backend are "last guard". But yes, these messages will be sent to client if client misses these rules, and client should present it. The guid could be skipped, but can also be important if end user report error

File: src/Equinor.ProCoSys.Completion.Command/PunchItemCommands/DeletePunchItemAttachment/DeletePunchItemAttachmentCommandValidator.cs (hunk 2 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.Command/PunchItemCommands/DeletePunchItemAttachment/DeletePunchItemAttachmentCommandValidator.cs:30]: I think so yes. See answer on comment above.

After a Punch is verified, the punch is "locked" for further editing (unless you are the person who verified it)
The UI should disable all functionality for further editing inclusive working with attachments (upload, delete, change)

**Review by trondlindanger — APPROVED** [review]

**Pull request merged**